File: IAPHelper.m
//
//  IAPHelper.m
//  BookClient
//
//  Created by Leo on 2016/12/18.
//  Copyright © 2016年 Leo. All rights reserved.
//

#import "IAPHelper.h"
#import "GTMBase64.h"

#ifdef DEBUG
#define checkURL @"https://sandbox.itunes.apple.com/verifyReceipt"
#else
#define checkURL @"https://buy.itunes.apple.com/verifyReceipt"
#endif
#define AppStoreInfoLocalFilePath [NSString stringWithFormat:@"%@/%@/", [NSSearchPathForDirectoriesInDomains(NSDocumentDirectory, NSUserDomainMask, YES) lastObject],@"EACEF35FE363A75A"]

@interface IAPHelper () <SKProductsRequestDelegate, SKPaymentTransactionObserver>

@end

@implementation IAPHelper {
    void(^_success)();
    void(^_fail)(NSError *error);
}

+ (instancetype)sharedHelper {
    static IAPHelper *helper;
    static dispatch_once_t onceToken;
    dispatch_once(&onceToken, ^{
        helper = [[self alloc] init];
    });
    return helper;
}

- (void)dealloc {
    [[SKPaymentQueue defaultQueue] removeTransactionObserver:self];
}

- (instancetype)init {
    if (self = [super init]) {
        [[SKPaymentQueue defaultQueue] addTransactionObserver:self];
    }
    return self;
}

- (void)payForProductWithProductID:(NSString *)productID success:(void(^)())success fail:(void(^)(NSError *error))fail {
    if (![SKPaymentQueue canMakePayments]) {             //替换项目时此处需修改
        NSError *error = [[NSError alloc] initWithDomain:@"audio.book.cast.review" code:0 userInfo:@{@"error": @"cannot make payments"}];
        fail(error);
        return;
    }
    _success = success;
    _fail = fail;
    NSSet *IDs = [NSSet setWithObject:productID];
    [self requestProductsWithProductIDs:IDs];
}

- (void)requestProductsWithProductIDs:(NSSet *)productIDs {
    SKProductsRequest *request = [[SKProductsRequest alloc] initWithProductIdentifiers:productIDs];
    request.delegate = self;
    [request start];
}

#pragma mark - SKProductsRequestDelegate
- (void)productsRequest:(SKProductsRequest *)request didReceiveResponse:(SKProductsResponse *)response {
    SKProduct *product = response.products.firstObject;
    if (product == nil) {
        NSError *error = [[NSError alloc] initWithDomain:@"audio.book.cast.review" code:0 userInfo:@{@"error": @"requested product not found"}];
        if (_fail) {
            _fail(error);
        }
        return;
    }
    SKPayment *pay = [SKPayment paymentWithProduct:product];
    [[SKPaymentQueue defaultQueue] addPayment:pay];
}

- (void)request:(SKRequest *)request didFailWithError:(NSError *)error {
    if (_fail) {
        _fail(error);
    }
}
#pragma Restored

+(void)stateRestored{

[[SKPaymentQueue defaultQueue] restoreCompletedTransactions];

}

-(void)paymentQueueRestoreCompletedTransactionsFinished:(SKPaymentQueue *)queue{

   NSMutableArray * purchasedItemIDs = [[NSMutableArray alloc] init];
    if (purchasedItemIDs.count == 0) {
        NSNotification *notification =[NSNotification notificationWithName:@"paymentfailedtongzhi" object:nil userInfo:nil];
        [[NSNotificationCenter defaultCenter] postNotification:notification];
    }
    
    NSLog(@"received restored transactions: %lu", (unsigned long)queue.transactions.count);
    for (SKPaymentTransaction *transaction in queue.transactions)
    {
        NSString *productID = transaction.payment.productIdentifier;
        [purchasedItemIDs addObject:productID];
        NSLog(@"%@",purchasedItemIDs);
    }

}

#pragma mark - Observer
- (void)paymentQueue:(SKPaymentQueue *)queue updatedTransactions:(NSArray<SKPaymentTransaction *> *)transactions {
    NSString *errorReason = @"";
    for (SKPaymentTransaction *transaction in transactions) {
        switch (transaction.transactionState) {
            case SKPaymentTransactionStatePurchased:
//                 获取并加密支付凭证
//                transaction.transactionReceipt
                self.receipt = [GTMBase64 stringByEncodingData:[NSData dataWithContentsOfURL:[[NSBundle mainBundle] appStoreReceiptURL]]];
                [self saveReceipt];
                [queue finishTransaction:transaction];
                if (_success) {
                    _success();
                }
                _success = nil;
                _fail = nil;
                
            {
                
                NSUserDefaults *defaults = [NSUserDefaults standardUserDefaults];
                [defaults setBool:YES forKey:@"paySuccess"];
                [defaults synchronize];
                

                NSNotification *notification =[NSNotification notificationWithName:@"tongzhi" object:nil userInfo:nil];
                [[NSNotificationCenter defaultCenter] postNotification:notification];
            }
                return;
            case SKPaymentTransactionStatePurchasing:
                return;
            case SKPaymentTransactionStateFailed:
                errorReason = @"payment failed";
            {
                NSNotification *notification =[NSNotification notificationWithName:@"paymentfailedtongzhi" object:nil userInfo:nil];
                [[NSNotificationCenter defaultCenter] postNotification:notification];
            
            }
                [queue finishTransaction:transaction];
                break;
            case SKPaymentTransactionStateRestored:
                
            {
                NSUserDefaults *defaults = [NSUserDefaults standardUserDefaults];
                [defaults setBool:YES forKey:@"paySuccess"];
                [defaults synchronize];
            
                [queue finishTransaction:transaction];
                
                errorReason = @"already pay for this product";
                NSNotification *notification =[NSNotification notificationWithName:@"tongzhi" object:nil userInfo:nil];
                [[NSNotificationCenter defaultCenter] postNotification:notification];

            }
                
                break;
            case SKPaymentTransactionStateDeferred:
                errorReason = @"payment state deferred";
                [queue finishTransaction:transaction];
                break;
            default:
                break;
        }
    }
    NSError *error = [[NSError alloc] initWithDomain:@"audio.book.cast.review" code:0 userInfo:@{@"error": errorReason}];
    if (_fail) _fail(error);
    _success = nil;
    _fail = nil;
}

//持久化存储用户购买凭证
-(void)saveReceipt {
    NSFileManager* fm = [NSFileManager defaultManager];
    NSString *documentPath = [NSSearchPathForDirectoriesInDomains(NSDocumentDirectory, NSUserDomainMask, YES) lastObject];
    NSString *dicPath = [documentPath stringByAppendingPathComponent:@"EACEF35FE363A75A"];
    if (![fm fileExistsAtPath:dicPath]) {
        [fm createDirectoryAtPath:dicPath withIntermediateDirectories:YES attributes:nil error:nil];
    }
    NSString *savedPath = [dicPath stringByAppendingString:@"/localTransactionReceipt.plist"];
    if (![fm fileExistsAtPath:savedPath]) {
        BOOL rrr = [fm createFileAtPath:savedPath contents:nil attributes:nil];
        if (rrr) {
            NSLog(@"%@",savedPath);
            NSLog(@"创建文件成功");
        } else {
            NSLog(@"创建文件失败");
        }
    }
    NSDictionary *dic =[ NSDictionary dictionaryWithObjectsAndKeys:
                        self.receipt,                           @"transactionReceipt",
                        nil];
    BOOL u =  [dic writeToFile:savedPath atomically:YES];
    NSLog(@"%d",u);
}
// App进入验证
- (void)sendFailedIapFiles{
    NSFileManager *fileManager = [NSFileManager defaultManager];
    NSError *error = nil;
    //搜索该目录下的所有文件和目录
    NSArray *cacheFileNameArray = [fileManager subpathsAtPath:AppStoreInfoLocalFilePath];
    if (error == nil)
    {
        for (NSString *name in cacheFileNameArray)
        {
            if ([name hasSuffix:@".plist"])//如果有plist后缀的文件，说明就是存储的购买凭证
            {
                NSString *filePath = [NSString stringWithFormat:@"%@/%@", AppStoreInfoLocalFilePath, name];
                [self sendAppStoreRequestBuyPlist:filePath];
            }
        }
    }
}

//获取购买的凭证
-(void)sendAppStoreRequestBuyPlist:(NSString *)plistPath
{
    NSDictionary *dic = [NSDictionary dictionaryWithContentsOfFile:plistPath];
    NSURL *url = [NSURL URLWithString:checkURL];
    NSMutableURLRequest *request = [NSMutableURLRequest requestWithURL:url cachePolicy:NSURLRequestUseProtocolCachePolicy timeoutInterval:20.0f];
    request.HTTPMethod = @"POST";
    NSString *encodeStr = [dic objectForKey:@"transactionReceipt"];
    NSString *payload = [NSString stringWithFormat:@"{\"receipt-data\" : \"%@\", \"password\" : \"%@\"}", encodeStr, @"97867822ceb24aa6b5b86902c00312d9"];
    NSData *payloadData = [payload dataUsingEncoding:NSUTF8StringEncoding];
    request.HTTPBody = payloadData;
    // 提交验证请求，并获得官方的验证JSON结果
    NSData *result = [NSURLConnection sendSynchronousRequest:request returningResponse:nil error:nil];
    
    if (result != nil) {
        NSDictionary *dict = [NSJSONSerialization JSONObjectWithData:result
                                                             options:NSJSONReadingAllowFragments error:nil];
        if (dict != nil) {
            // 验证成功,通知
            NSArray *receiptInfo = dict[@"latest_receipt_info"];
            NSDictionary *lastReceiptInfo = receiptInfo.lastObject;
            //本地存储第一次订阅时间
            NSDictionary * firstReceiptInfo = receiptInfo.firstObject;
            NSString * purchase = firstReceiptInfo[@"purchase_date_ms"];
            NSUserDefaults *defaults = [NSUserDefaults standardUserDefaults];
            [defaults setObject:purchase forKey:@"purchase_date_ms"];
            [defaults synchronize];
            if (lastReceiptInfo != nil) {
                //到期时间
                NSString *expires = lastReceiptInfo[@"expires_date_ms"];
                //第一次订阅时间
                long long longExpires = [expires longLongValue];
                NSTimeInterval current = [[NSDate date] timeIntervalSince1970];
                long long longCurrent = (long long)current;
                long diff = (long)(longCurrent * 1000 - longExpires );
                if( diff > 0){
                    if (self.payVerificationDelegate && [self.payVerificationDelegate respondsToSelector:@selector(payVerificationfailed)]) {
                        [self.payVerificationDelegate payVerificationfailed];
                    }
                } else {
                    //未过期  本地存储信息 （expires_date_ms：到期日的时间戳）
                    NSUserDefaults *defaults = [NSUserDefaults standardUserDefaults];
                    [defaults setObject:expires forKey:@"expires_date_ms"];
                    [defaults setObject:purchase forKey:@"purchase_date_ms"];
                    [defaults synchronize];
                    
                    if (self.payVerificationDelegate && [self.payVerificationDelegate respondsToSelector:@selector(payVerificationSuccess)]) {
                        [self.payVerificationDelegate payVerificationSuccess];
                    }
                }
            }
        }else{
            if (self.payVerificationDelegate && [self.payVerificationDelegate respondsToSelector:@selector(payVerificationSuccess)]) {
                [self.payVerificationDelegate payVerificationSuccess];
            }
        }
    }
}


@end

File: swift_pay/IAPHelper.swift
//
//  IAPHelper.swift
//  iMessageKeyboard
//
//  Created by 杰刘 on 2018/3/29.
//  Copyright © 2018年 刘杰. All rights reserved.
//

import UIKit
import StoreKit

protocol IAPDelegate:class {
    func paysuccess()
    func payError()
    func payrestoreError()
    func payrestoresuccess()
}

class IAPHelper: NSObject {
    
    weak var delegate: IAPDelegate?
    
    override init() {
        super.init()
    }
    deinit{
        SKPaymentQueue.default().remove(self as SKPaymentTransactionObserver)
    }
    
    //开始请求订阅
    func payForProductWithProductID(productID:String)  {
        guard productID.isEmpty == false else {
            return
        }
        //判断当前的支付环境, 是否可以支付
        guard SKPaymentQueue.canMakePayments() else {
            //错误的支付环境
            self.delegate?.payError()
            return
        }
        //填写商品的id
        let ids = NSSet.init(object: productID)
        let request: SKProductsRequest = SKProductsRequest.init(productIdentifiers: ids as! Set<String>)//上面的商品还要到苹果服务器进行验证, 看下哪些商品是可以真正被销售的（创建一个商品请求并设置请求的代理，由代理告知结果）
        request.delegate = self//设置代理, 接收可以被销售的商品列表数据
        request.start()//转到extension ViewController: SKProductsRequestDelegate {
    }
    //持久化存储用户购买凭证
    func saveReceipt(receipt:String) {
        
        let fm = FileManager.default
        let documnetPath = NSSearchPathForDirectoriesInDomains(FileManager.SearchPathDirectory.documentDirectory, FileManager.SearchPathDomainMask.userDomainMask, true)[0] as NSString
        let dicPath = documnetPath.appendingPathComponent("EACEF35FE363A75A")
        if !fm.fileExists(atPath: dicPath) {
            do {
                try fm.createDirectory(atPath: dicPath, withIntermediateDirectories: true, attributes: nil)
            }catch {
            }
        }
        let savedPath = dicPath.appending("/localTransactionReceipt.plist")
        if !fm.fileExists(atPath: dicPath) {
            
            let rrr = fm.createFile(atPath: savedPath, contents: nil, attributes: nil)
            if rrr {
                print("创建文件成功")
            }else{
                print("创建文件失败")
            }
        }
        let dict = NSDictionary.init(object: receipt, forKey: "transactionReceipt" as NSCopying)
        let u = dict.write(toFile: savedPath, atomically: true)
        print("写入是否成功 \(u)")
    }
    //恢复订阅
    func restore() {
        //恢复购买
        SKPaymentQueue.default().restoreCompletedTransactions()
        SKPaymentQueue.default().add(self as SKPaymentTransactionObserver)
    }
    func paymentQueueRestoreCompletedTransactionsFinished(_ queue: SKPaymentQueue) {
        //写入本地
        let receipt = LJBase64.base64Encoding(plainData: try! Data.init(contentsOf: Bundle.main.appStoreReceiptURL!))
        if receipt.isEmpty == false {
            self.saveReceipt(receipt: receipt)
        }
        self.delegate?.payrestoresuccess()
        
    }
    func paymentQueue(_ queue: SKPaymentQueue, restoreCompletedTransactionsFailedWithError error: Error) {
        self.delegate?.payrestoreError()
    }
    
}

extension IAPHelper:SKPaymentTransactionObserver{
    
    // 当交易队列里面添加的每一笔交易状态发生变化的时候调用
    func paymentQueue(_ queue: SKPaymentQueue, updatedTransactions transactions: [SKPaymentTransaction]) {
        for transaction in transactions {
            
            print("订阅失败信息  \(String(describing: transaction.error))")
            switch transaction.transactionState {
            case .deferred:
                print("延迟处理")
            case .failed:
                print("支付失败")
                queue.finishTransaction(transaction)
                self.delegate?.payError()
            case .purchased:
                print("支付成功")
                //写入本地
                let receipt = LJBase64.base64Encoding(plainData: try! Data.init(contentsOf: Bundle.main.appStoreReceiptURL!))
                if receipt.isEmpty == false {
                    self.saveReceipt(receipt: receipt)
                }
                self.delegate?.paysuccess()
                queue.finishTransaction(transaction)
            case .purchasing:
                print("正在支付")
            case .restored:
                print("恢复购买")
                queue.finishTransaction(transaction)
            }
        }
    }
    
}

extension IAPHelper:SKProductsRequestDelegate{
    func productsRequest(_ request: SKProductsRequest, didReceive response: SKProductsResponse) {
        //当请求完毕之后, 从苹果服务器获取到数据之后调用
        print(response.products)
        print(response.invalidProductIdentifiers)
        let product = response.products.first
        
        if product == nil {
            //错误的商品信息
            self.delegate?.payError()
            return
        }
        
        for product in response.products {
            // 激活了对应的销售操作按钮，相当于商店的商品上架允许销售
            print("=======Product id=======\(product.productIdentifier)")
            print("===产品标题 ==========\(product.localizedTitle)")
            print("====产品描述信息==========\(product.localizedDescription)")
            print("=====价格: =========\(product.price)")
        }
        let payment = SKPayment(product: product!)
        SKPaymentQueue.default().add(payment)//添加到支付队列
        SKPaymentQueue.default().add(self as SKPaymentTransactionObserver)//监听交易状态
    }
    
    func request(_ request: SKRequest, didFailWithError error: Error) {
        //验证失败
        self.delegate?.payError()
    }
    func requestDidFinish(_ request: SKRequest) {
        print("返回信息结束调用")
    }
}
